### PR TITLE
Fix to intermediate download of a resource's content

### PR DESCRIPTION
**Pull request by jsmatias**
I replaced the `request.get` in the code by `RedirectResponse`. This way the aiod redirects the call to the specific URL to download the content of a resource. Then, the download will be handled on the client side instead of downloading on the server side and passing it to the client.

### CODE REVIEW

File: src/routers/resource_ai_asset_router.py [hunk 1]
```diff
@@ -82,25 +81,17 @@ def get_resource_content(
...
+            encoding_format = distributions[distribution_idx].encoding_format
+            filename = distributions[distribution_idx].name
+
+            headers = {
+                "Content-Disposition": ("attachment; " f"filename={filename or url.split('/')[-1]}")
+            }
+            if encoding_format:
+                headers["Content-Type"] = encoding_format
+
+            return RedirectResponse(url=url, status_code=status.HTTP_303_SEE_OTHER, headers=headers)
```
Comment (josvandervelde) [inline src/routers/resource_ai_asset_router.py:94]: The headers are not doing anything. It's explained quite well here: https://stackoverflow.com/questions/7583461/redirect-to-page-and-send-custom-http-headers/41218304#41218304 

I tested this using a Zenodo upload. It has `"name": "croissant(11).json"` in the `distribution`. It should get a "Content-Disposition" header with `"attachment; filename=croissant(11).json"`. But it get's an `inline` header if I GET http://localhost/datasets/v1/1/content, which is what you get from Zenodo.

I don't see any possibility to "force" the headers. We redirect the user, and then Zenodo (or whatever platform) controls the response. If this is true, we should probably remove the headers..

**Review by josvandervelde — CHANGES_REQUESTED** [review]
Nice idea, unfortunately I'm afraid it's not working as we'd like, see comments

File: src/routers/resource_ai_asset_router.py (hunk 1 above)
Comment (jsmatias) [reply src/routers/resource_ai_asset_router.py]: Indeed, I removed the headers in this case.

File: src/routers/resource_ai_asset_router.py [hunk 2]
```diff
@@ -82,25 +81,17 @@ def get_resource_content(
...
-                return Response(content=content, headers=headers)
-
-            except Exception as exc:
-                raise _wrap_as_http_exception(exc)
+            url = distributions[distribution_idx].content_url
+            encoding_format = distributions[distribution_idx].encoding_format
+            filename = distributions[distribution_idx].name
+
+            headers = {
+                "Content-Disposition": ("attachment; " f"filename={filename or url.split('/')[-1]}")
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply src/routers/resource_ai_asset_router.py]: Nice spotted! The strings were displaced in two lines and I missed it when I placed them in a single line.

File: src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py [hunk 3]
```diff
@@ -93,24 +45,23 @@ def test_ai_asset_has_endpoints(
...
 
-    with responses.RequestsMock() as mocked_requests:
-        mock_response1(mocked_requests)
-
-        response = client.get(default_endpoint)
-        assert response.status_code == status.HTTP_200_OK, response.json()
-        response0 = client.get(default_endpoint + "/0")
-        assert response0.status_code == status.HTTP_200_OK, response0.json()
+    response = client.get(default_endpoint, allow_redirects=False)
+    assert response.status_code == status.HTTP_303_SEE_OTHER, response.status_code
```
Comment (josvandervelde) [inline src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py:56]: It's probably good to test the `response.headers["location"]` , at least once

**Review by josvandervelde — APPROVED** [review]
Looks good to me! One minor comment.

It's a shame that's it not possible to keep the headers (I really like the earlier headers) but this is better.

File: src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py (hunk 3 above)
Comment (jsmatias) [reply src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py:56]: This is already included on test functions:
- `test_endpoints_when_single_distribution`,
- `test_endpoints_when_two_distributions`

**Pull request merged**